### PR TITLE
Another bucket sort

### PR DESCRIPTION
**Pull request by ikawrakow**
We now have 3 PR's related to sorting logits with the goal to speedup top-k sampling:
* PR #5085 - speed up via usage of `std::nth_element` before `std::partial_sort`
* PR #5101, which introduces a bucket sort algorithm
* This PR, which provides a better (faster) implementation of the bucket sort proposed in #5101 

The table shows a comparison between master and these 3 PR's as a function of `top_k`. Tests run on a Ryzen-5975WX + RTX--4080 with Ubuntu 22.04 and GCC 11.4.0.

| --top-k | t/s master | t/s (PR 5085) | t/s PR 5101 | t/s this PR | Speedup this PR |
|---------|------------|-----------------|---------------------|-----------------|---------------------|
|      100 |       7758 |            7758 |   3084 | 7758 | 1.000 |
|      200 |       6796 |            6796 |   3076 | 7121 |  1.048 |
|      500 |       4848 |            4848 |    2905 | 6733 | 1.389 |
|    1000 |       3353 |            3353 |    2872 | 6075 | 1.812 |
|    2000 |       2123 |            2123 |    2592 | 5006 | 2.358 |
|    4000 |       1288 |            2000 |    2246 | 3627 | 2.816 |
|    8000 |         783 |            1438 |    1718 | 2292 | 2.927 |
|   16000 |        508 |             921 |     1143 | 1326 | 2.611 |
|   31780 |        397 |             558 |       675 | 734  | 1.848 |
|   32000 |        559 |             559 |       703 | 773 |  1.384 |

### CODE REVIEW

**Comment by cmp-nct** [comment]
If this continues we'll soon see a GPU tensor-core sorting kernel beating this one again :)
It's quite amazing to see how much performance can be gained.
On the other hand, I think pre-filtering K to a more meaningful value is still the best way to go in terms of practicality, if huge-top-k values are the future.

**Comment by ikawrakow** [comment]
> If this continues we'll soon see a GPU tensor-core sorting kernel beating this one again :) It's quite amazing to see how much performance can be gained. On the other hand, I think pre-filtering K to a more meaningful value is still the best way to go in terms of practicality, if huge-top-k values are the future.

Well, when the GPU beats this, there is still room for improvement. One can easily shave off another 10% or so from the time by having a top_k sampler instance that has the buffers pre-allocated, so one doesn't need to do memory allocations on each invocation of `llama_sample_top_k`. One can vectorize at least the 1st loop. Etc.

More seriously, I do agree with you that if the usage if large top_k becomes standard practice, it is better to prefilter the logits in some way.

**Comment by JohannesGaessler** [comment]
PR for a min_p implementation that works on unsorted tokens: https://github.com/ggerganov/llama.cpp/pull/5115 .

**Review by JohannesGaessler — APPROVED** [review]
LGTM; the code has become more difficult to understand but in my opinion speed is more important unless the code is irrelevant for performance. But maybe we should wait for the opinion of another dev just to be sure.

I can confirm that the performance is better than both master and my bucket sort PR:

bucket_sort

Good job!
